### PR TITLE
config: add formatlist

### PR DESCRIPTION
**Pull request by josharian**
With this PR, format now distributes over lists. Please see the commit comment and doc changes in f2fb9e7 for motivation and details.

### CODE REVIEW

**Comment by radeksimko** [comment]
I like the idea and understand the motivation, but I feel that reusing the current `format` function is making that original function way too magical. I'd personally rather see a separate function handling this.

The reason is that I would never expect `list` to be output from a function that's called `format`.

**Comment by josharian** [comment]
Would it help if `format` stayed as it was, and the newly-empowered function was called `formatlist`? `formatlist` could return an error if none of the params was a list.

FWIW, I experimented with a much less powerful version of this (https://github.com/josharian/terraform/commit/ed79154b009f7dbece60a4cfe6702467ce994e90), but concluded that it was helpful to be able to mix lists and scalars and to have multiple lists.

**Comment by josharian** [comment]
Ok, I've moved list-oriented formatting to `formatlist`. Please take another look. (Also, let me know whether you'd like me to squash the history before merging. Leaving as-is for now because GitHub PRs don't handle rebasing gracefully, grrrr.)

File: config/interpolate_walk.go [hunk 1]
```diff
@@ -11,8 +11,7 @@ import (
 )
 
 // InterpSplitDelim is the delimeter that is looked for to split when
-// it is returned. This is a comma right now but should eventually become
-// a value that a user is very unlikely to use (such as UUID).
+// it is returned.
```
Comment (radeksimko) [inline config/interpolate_walk.go]: Good spot.

**Comment by radeksimko** [comment]
@josharian This looks good to me, so unless another maintainer jumps in here with a reason for not merging this, I will merge it.

The last thing I'd ask you to do before merging though is to squash your commits.

**Comment by josharian** [comment]
@radeksimko thanks. Squashed down to two commits and rebased.

File: config/interpolate_funcs.go [hunk 2]
```diff
@@ -88,6 +90,69 @@ func interpolationFuncFormat() ast.Function {
...
+					n = len(parts)
+					continue
+				}
+				if n != len(parts) {
+					return nil, fmt.Errorf("format: mismatched list lengths: %d != %d", n, len(parts))
+				}
+			}
+
+			if n == 0 {
+				return nil, errors.New("no lists in arguments to formatlist")
```
Comment (radeksimko) [inline config/interpolate_funcs.go]: Sorry I didn't notice this before, but why exactly don't you use `fmt.Errorf()` here?

Comment (radeksimko) [inline config/interpolate_funcs.go]: Stupid question... the answer is because you don't need any formatting :open_hands:

**Pull request merged**